fabric-enchantment-api-v1/src/main/java/net/fabricmc/fabric/api/enchantment/v1/EnchantmentEvents.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.enchantment.v1;
+
+import net.minecraft.enchantment.Enchantment;
+import net.minecraft.item.ItemStack;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.api.util.TriState;
+
+/**
+ * Events to intercept the vanilla enchantment logic.
+ *
+ * @author Vaerian (vaeriann@gmail.com or @Vaerian on GitHub).
+ *
+ * <p>Please contact the author, Vaerian, at the email or GitHub profile listed above
+ * with any questions surrounding implementation choices, functionality, or updating
+ * newer versions of the game.</p>
+ */
+public class EnchantmentEvents {
+	/**
+	 * Overrides whether or not the given enchantment should be applied to the given item
+	 * only when enchanted using an anvil.
+	 *
+	 * <p>Callbacks are evaluated on a first come, first serve basis where callbacks
+	 * registered earlier will have functional priority over ones registered later.</p>
+	 */
+	public static final Event<AcceptEnchantment> ACCEPT_APPLICATION = EventFactory.createArrayBacked(AcceptEnchantment.class, (callbacks) -> (enchantment, stack) -> {
+		for (AcceptEnchantment callback : callbacks) {
+			TriState result = callback.shouldAccept(enchantment, stack);
+
+			if (result != TriState.DEFAULT) {
+				return result;
+			}
+		}
+
+		return TriState.DEFAULT;
+	});
+
+	/**
+	 * Overrides whether or not the given enchantment should be applied to the given item
+	 * only when enchanted using an enchantment table.
+	 *
+	 * <p>Callbacks are evaluated on a first come, first serve basis where callbacks
+	 * registered earlier will have functional priority over ones registered later.</p>
+	 */
+	public static final Event<AcceptEnchantment> ACCEPT_ENCHANTMENT = EventFactory.createArrayBacked(AcceptEnchantment.class, (callbacks) -> (enchantment, stack) -> {
+		for (AcceptEnchantment callback : callbacks) {
+			TriState result = callback.shouldAccept(enchantment, stack);
+
+			if (result != TriState.DEFAULT) {
+				return result;
+			}
+		}
+
+		return TriState.DEFAULT;
+	});
+
+	/**
+	 * Simple helper method to add an enchantment callback to both the enchantment and application
+	 * events.
+	 *
+	 * @param callback The enchantment callback to add to both enchantment events.
+	 */
+	public static void registerAll(AcceptEnchantment callback) {
```

Comment:
I believe the standards for events here is that Accept and accept application would be two separate functional interfaces.

Reply:
How would you propose handling a method such as this one if they were to become two separate interfaces? I was thinking maybe casting a BiFunction, or defining a parent functional interface to cast from.